src/seat/input_method_v3.rs

```diff
@@ -0,0 +1,711 @@
+    pub fn get_positioner<State>(&self, qh: &QueueHandle<State>) -> PopupPositioner
+    where
+        State: Dispatch<XxInputPopupPositionerV1, PositionerData, State> + 'static,
+    {
+        PopupPositioner(self.manager.get_positioner(qh, PositionerData))
+    }
+}
+
+impl<D> Dispatch<xx_input_method_manager_v2::XxInputMethodManagerV2, GlobalData, D>
+    for InputMethodManager
+where
+    D: Dispatch<xx_input_method_manager_v2::XxInputMethodManagerV2, GlobalData>
+        + InputMethodHandler,
+{
+    fn event(
+        _data: &mut D,
+        _manager: &xx_input_method_manager_v2::XxInputMethodManagerV2,
+        _event: xx_input_method_manager_v2::Event,
+        _: &GlobalData,
+        _conn: &Connection,
+        _qh: &QueueHandle<D>,
+    ) {
+        unreachable!()
+    }
+}
+
+/// A trivial wrapper for an [`XxInputPopupPositionerV1`].
+///
+/// This wrapper calls [`destroy`][XxInputPopupPositionerV1::destroy] on the contained
+/// positioner when it is dropped.
+#[derive(Debug)]
+pub struct PopupPositioner(XxInputPopupPositionerV1);
+
+impl Deref for PopupPositioner {
+    type Target = XxInputPopupPositionerV1;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Drop for PopupPositioner {
+    fn drop(&mut self) {
+        self.0.destroy()
+    }
+}
+
+impl<D> Dispatch<XxInputPopupPositionerV1, PositionerData, D> for PopupPositioner
+where
+    D: Dispatch<XxInputPopupPositionerV1, PositionerData> + InputMethodHandler,
+{
+    fn event(
+        _data: &mut D,
+        _manager: &XxInputPopupPositionerV1,
+        _event: xx_input_popup_positioner_v1::Event,
+        _: &PositionerData,
+        _conn: &Connection,
+        _qh: &QueueHandle<D>,
+    ) {
+        unreachable!("Positioner has no events")
+    }
+}
+
+#[derive(Debug)]
+pub struct PositionerData;
+
+#[derive(Debug)]
+pub struct InputMethod {
+    input_method: XxInputMethodV1,
+}
+
+impl InputMethod {
+    pub fn input_method(&self) -> &XxInputMethodV1 {
+        &self.input_method
+    }
+
+    pub fn set_preedit_string(&self, text: String, cursor: CursorPosition) {
+        // TODO: should this enforce indices on codepoint boundaries?
+        let (start, end) = match cursor {
+            CursorPosition::Hidden => (-1, -1),
+            CursorPosition::Visible { start, end } => (
+                // This happens only for cursor values in the upper usize range.
+                // Such values are most likely bugs already,
+                // so it's not a problem if one of the cursors weirdly lands at 0 sometimes.
+                start.try_into().unwrap_or(0),
+                end.try_into().unwrap_or(0),
+            ),
+        };
+        self.input_method.set_preedit_string(text, start, end)
+    }
+
+    pub fn commit_string(&self, text: String) {
+        self.input_method.commit_string(text)
+    }
+
+    pub fn delete_surrounding_text(&self, before_length: u32, after_length: u32) {
+        // TODO: this has 2 separate behaviours:
```

Comment:
I'm not sure that I understand the difference in the protocol. The preedit text is just removed in the case that it is present, and otherwise it works the same, right?